File: Leetcode/Letter Combinations of a Phone Number.cc
class Solution {
public:
    vector<string> letterCombinations(string digits) {
        vector<string> res;
        string t;
        dfs(digits, 0, t, res);
        return res;
    }
    
    void dfs(const string& digits, int p, string t, vector<string>& res) {
        if (p == digits.size()) {
            res.push_back(t);
            return;
        }
        
        int now = digits[p] - '0';
        for (int i = 0; i < v[now].size(); i++) {
            dfs(digits, p + 1, t + v[now][i], res);
        }
    }
    
    Solution(){
        v[2] = "abc";
        v[3] = "def";
        v[4] = "ghi";
        v[5] = "jkl";
        v[6] = "mno";
        v[7] = "pqrs";
        v[8] = "tuv";
        v[9] = "wxyz";
    }
    
    string v[10];
};

File: LintCode/python/Letter Combinations of a Phone Number.py
class Solution:
    # @param {string} digits A digital string
    # @return {string[]} all posible letter combinations
    def letterCombinations(self, digits):
        self.keyboard = {
            2: 'abc',
            3: 'def',
            4: 'ghi',
            5: 'jkl',
            6: 'mno',
            7: 'pqrs',
            8: 'tuv',
            9: 'wxyz'
        }
        self.res = []
        self.buffer = []
        self.dfs(digits)
        return self.res

    def dfs(self, digits):
        if not digits:
            s = ''.join(self.buffer)
            if s:
                self.res.append(s)
            return
        d = int(digits[0])
        for item in self.keyboard[d]:
            self.buffer.append(item)
            self.dfs(digits[1:])
            self.buffer.pop()
